### PR TITLE
C++: Expose memory backpressure to users

### PR DESCRIPTION
**Pull request by senior7515**

### CODE REVIEW

File: src/core/rpc_recv_context.cc [hunk 1]
```diff
@@ -112,12 +103,24 @@ rpc_recv_context::parse_header(rpc_connection *conn) {
...
+      if (hdr.compression() > rpc::compression_flags_MAX) {
+        LOG_ERROR("Compression out of range", hdr);
+        return seastar::make_ready_future<ret_type>(stdx::nullopt);
+      }
+      if (hdr.checksum() <= 0) {
+        LOG_ERROR("checksum is empty");
+        return seastar::make_ready_future<ret_type>(stdx::nullopt);
+      }
+      if (hdr.meta() <= 0) {
+        LOG_ERROR("meta is empty");
```
Comment (senior7515) [inline src/core/rpc_recv_context.cc]: DLOG*

File: src/core/rpc_server.cc [hunk 2]
```diff
@@ -142,51 +143,59 @@ rpc_server::stop() {
...
+              }
+              //
+              // Launch the actual processing on a background
+              // future
+              //
+              dispatch_rpc(conn, std::move(maybe_payload.value()))
+                .finally([m = hist_->auto_measure(), conn, payload_size] {
+                  conn->limits()->resources_available.signal(payload_size);
+                })
+                .handle_exception(
```
Comment (senior7515) [inline src/core/rpc_server.cc]: not needed.

File: src/core/rpc_client.cc [hunk 3]
```diff
@@ -167,34 +175,67 @@ process_body_for_self(stdx::optional<rpc_recv_context> opt, auto self) {
...
+           [this]() {
+             auto timeout = conn->limits->max_body_parsing_duration;
+             seastar::timer<> body_timeout;
+             body_timeout.set_callback([timeout, this] {
+               LOG_ERROR(
+                 "rpc_client exceeded timeout: {}ms",
+                 std::chrono::duration_cast<std::chrono::milliseconds>(timeout)
+                   .count());
+               conn->set_error("Connection body parsing exceeded timeout");
+               conn->socket.shutdown_input();
```
Comment (senior7515) [inline src/core/rpc_client.cc]: clients are different than servers. One client conn can timeout.

File: src/core/rpc_client.cc [hunk 4]
```diff
@@ -167,34 +175,67 @@ process_body_for_self(stdx::optional<rpc_recv_context> opt, auto self) {
...
-                   });
+           [this]() {
+             auto timeout = conn->limits->max_body_parsing_duration;
+             seastar::timer<> body_timeout;
+             body_timeout.set_callback([timeout, this] {
+               LOG_ERROR(
+                 "rpc_client exceeded timeout: {}ms",
+                 std::chrono::duration_cast<std::chrono::milliseconds>(timeout)
+                   .count());
+               conn->set_error("Connection body parsing exceeded timeout");
```
Comment (senior7515) [inline src/core/rpc_client.cc]: should instead `fullfill_timeoutsession(uint16_t) which can just pass in nullopt ? `

need to brainstorm.

**Comment by Lwelch45** [comment]
I was just going to raise an issue related to connection disconnects. It seems that smf operates under the assumption that the client maintains a long lived connection to the server. Would the user handle reconnects to the server in the case of network blips?
For context im looking to implement raft on top of smf which would require the servers to be up and connected to each other for fairly long periods of time

**Comment by senior7515** [comment]
@Lwelch45  good stuff. You would be the third user implementing raft *at this moment* on top of smf. 

@Lwelch45  - please open up a new issue. There has to be some brainstorm of what a reasonable API would look like for reconnects. Typically most frameworks - twitter's Finagle, new facebook-thrift w/ folly, etc, all delegate this to the user. What they provide is a a way to detect if the connection has failed (we do via `conn-is_valid()`). The user is then responsible for recreating a new pointer (via `GeneratedClient::make_shared(opts)` ) and trying again. 

However, the way this design came about, it is really simple in implementation since most of the code is codegen'ed. 

When you create a new ticket. we should list alternative designs for reconnections. 

Maybe a nice filter api can do it. i.e.:

``` client.register_outgoing_filter < smf::retry_tcp_conn<3>>() ```

**Comment by senior7515** [comment]
@senior7515  - we need to fulfill an exceptional future on conn.close() to give user's the chance  to deal w/ exceptional returns from function calls rather than just silently doing just a tcp.connecttion aborted.

File: src/core/rpc_client.cc [hunk 5]
```diff
@@ -167,34 +175,67 @@ process_body_for_self(stdx::optional<rpc_recv_context> opt, auto self) {
...
+        // expected. no outstanding reqs
+        conn->disable();
+      }
+      return seastar::make_ready_future<>();
+    }
+    return rpc_recv_context::parse_payload(conn.get(), std::move(hdr.value()))
+      .then([this](stdx::optional<rpc_recv_context> opt) mutable {
+        if (!opt) {
+          LOG_ERROR("Could not parse response from server. Bad payload");
+          conn->set_error("Invalid payload");
```
[the remark this answers is not in the document]
Comment (senior7515) [reply src/core/rpc_client.cc]: done

**Pull request merged**